File: src/MostrarTextoGrisaceo CS/Form1.cs
//-----------------------------------------------------------------------------
// Mostrar efecto grisáceo (deshabilitado) en las cajas de texto    (18/Oct/20)
//
//
// (c) Guillermo (elGuille) Som, 2020
//-----------------------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

public partial class Form1 : Form
{

    #region La definición de los controles del formulario y void Main()

    /// <summary>
    /// Punto de entrada principal para la aplicación.
    /// </summary>
    [STAThread]
    static void Main()
    {
        Application.EnableVisualStyles();
        Application.SetCompatibleTextRenderingDefault(false);
        Application.Run(new Form1());
    }

    public Form1()
    {
        InitializeComponent();
    }

    /// <summary>
    /// Variable del diseñador necesaria.
    /// </summary>
    private System.ComponentModel.IContainer components = null;

    /// <summary>
    /// Limpiar los recursos que se estén usando.
    /// </summary>
    /// <param name="disposing">true si los recursos administrados se deben desechar; false en caso contrario.</param>
    protected override void Dispose(bool disposing)
    {
        if (disposing && (components != null))
        {
            components.Dispose();
        }
        base.Dispose(disposing);
    }


    /// <summary>
    /// Método necesario para admitir el Diseñador. No se puede modificar
    /// el contenido de este método con el editor de código.
    /// </summary>
    private void InitializeComponent()
    {
            this.txtTexto = new System.Windows.Forms.TextBox();
            this.btnAceptar = new System.Windows.Forms.Button();
            this.btnCancelar = new System.Windows.Forms.Button();
            this.labelInfo = new System.Windows.Forms.Label();
            this.SuspendLayout();
            // 
            // txtTexto
            // 
            this.txtTexto.Anchor = ((System.Windows.Forms.AnchorStyles)(((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Left) 
            | System.Windows.Forms.AnchorStyles.Right)));
            this.txtTexto.ForeColor = System.Drawing.SystemColors.GrayText;
            this.txtTexto.Location = new System.Drawing.Point(12, 14);
            this.txtTexto.Name = "txtTexto";
            this.txtTexto.Size = new System.Drawing.Size(295, 20);
            this.txtTexto.TabIndex = 4;
            this.txtTexto.TextChanged += new System.EventHandler(this.txtTexto_TextChanged);
            this.txtTexto.Enter += new System.EventHandler(this.txtTexto_Enter);
            this.txtTexto.Leave += new System.EventHandler(this.txtTexto_Leave);
            // 
            // btnAceptar
            // 
            this.btnAceptar.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Right)));
            this.btnAceptar.AutoSize = true;
            this.btnAceptar.Location = new System.Drawing.Point(313, 12);
            this.btnAceptar.Name = "btnAceptar";
            this.btnAceptar.Size = new System.Drawing.Size(75, 23);
            this.btnAceptar.TabIndex = 5;
            this.btnAceptar.Text = "Saludar";
            this.btnAceptar.Click += new System.EventHandler(this.btnAceptar_Click);
            // 
            // btnCancelar
            // 
            this.btnCancelar.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Bottom | System.Windows.Forms.AnchorStyles.Right)));
            this.btnCancelar.AutoSize = true;
            this.btnCancelar.DialogResult = System.Windows.Forms.DialogResult.Cancel;
            this.btnCancelar.Location = new System.Drawing.Point(313, 106);
            this.btnCancelar.Name = "btnCancelar";
            this.btnCancelar.Size = new System.Drawing.Size(75, 23);
            this.btnCancelar.TabIndex = 6;
            this.btnCancelar.Text = "Cerrar";
            this.btnCancelar.Click += new System.EventHandler(this.btnCancelar_Click);
            // 
            // labelInfo
            // 
            this.labelInfo.AutoSize = true;
            this.labelInfo.Location = new System.Drawing.Point(12, 50);
            this.labelInfo.Name = "labelInfo";
            this.labelInfo.Size = new System.Drawing.Size(129, 13);
            this.labelInfo.TabIndex = 7;
            this.labelInfo.Text = "<Escribe algo en el texto>";
            // 
            // Form1
            // 
            this.AutoScaleDimensions = new System.Drawing.SizeF(6F, 13F);
            this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font;
            this.ClientSize = new System.Drawing.Size(400, 141);
            this.Controls.Add(this.txtTexto);
            this.Controls.Add(this.btnAceptar);
            this.Controls.Add(this.btnCancelar);
            this.Controls.Add(this.labelInfo);
            this.Name = "Form1";
            this.Text = "Form1 C#";
            this.Load += new System.EventHandler(this.Form1_Load);
            this.ResumeLayout(false);
            this.PerformLayout();

    }

    private System.Windows.Forms.TextBox txtTexto;
    private System.Windows.Forms.Button btnAceptar;
    private System.Windows.Forms.Button btnCancelar;
    private System.Windows.Forms.Label labelInfo;

    #endregion

    private void Form1_Load(object sender, EventArgs e)
    {
        this.CenterToScreen();

        txtTexto.Text = textVacio;
        txtTexto.SelectionStart = 0; //txtTexto.Text.Length;
    }

    private void btnAceptar_Click(object sender, EventArgs e)
    {
        if( txtTexto.Text == textVacio)
            labelInfo.Text = "¡Hola amigo!";
        else
            labelInfo.Text = $"¡Hola {txtTexto.Text}!";
    }

    private void btnCancelar_Click(object sender, EventArgs e)
    {
        this.Close();
    }

    // 
    // Lo que necesitamos para hacer el efecto
    // 

    private const string textVacio = "<Escribe tu nombre>";
    private string textAnterior = textVacio;
    private bool inicializando;

    private void txtTexto_Enter(object sender, EventArgs e)
    {
        if (txtTexto.Text != textVacio)
            txtTexto.ForeColor = SystemColors.ControlText;
    }

    private void txtTexto_Leave(object sender, EventArgs e)
    {
        if (string.IsNullOrEmpty(this.txtTexto.Text))
        {
            this.txtTexto.ForeColor = SystemColors.GrayText;
            this.txtTexto.Text = textVacio;
        }
    }

    private void txtTexto_TextChanged(object sender, EventArgs e)
    {
        if (inicializando)
            return;
        if (txtTexto.Text == "" || txtTexto.Text == textVacio)
        {
            txtTexto.ForeColor = SystemColors.GrayText;
            inicializando = true;
            txtTexto.Text = textVacio;
            inicializando = false;
        }
        else
        {
            if (textAnterior == textVacio)
            {
                inicializando = true;
                txtTexto.Text = QuitarPredeterminado(txtTexto.Text, textVacio);
                inicializando = false;

                txtTexto.SelectionStart = txtTexto.Text.Length;
            }
            txtTexto.ForeColor = SystemColors.ControlText;
        }
        textAnterior = txtTexto.Text;
    }

    /// <summary>
    /// Quitar de una cadena un texto indicado (que será el predeterminado cuando está vacío).
    /// Por ejemplo si el texto grisáceo es Buscar... y
    /// se empezó a escribir en medio del texto (o en cualquier parte)
    /// BuscarL... se quitará Buscar... y se dejará L.
    /// </summary>
    /// <param name="texto">El texto en el que se hará la sustitución.</param>
    /// <param name="predeterminado">El texto a quitar.</param>
    /// <returns>Una cadena con el texto predeterminado quitado.</returns>
    /// <remarks>18/Oct/2020</remarks>
    private string QuitarPredeterminado(string texto, string predeterminado)
    {
        for (var i = 0; i < predeterminado.Length; i++)
        {
            var j = texto.IndexOf(predeterminado[i]);
            if (j == -1)
                continue;
            if (j == 0)
                texto = texto.Substring(j + 1);
            else
                texto = texto.Substring(0, j) + texto.Substring(j + 1);
        }
        return texto;
    }
}

File: src/MostrarTextoGrisaceo VB/Form1.vb
'------------------------------------------------------------------------------
' Mostrar efecto grisáceo (deshabilitado) en las cajas de texto     (18/Oct/20)
'
'
' (c) Guillermo (elGuille) Som, 2020
'------------------------------------------------------------------------------
Option Strict On
Option Infer On

Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
'Imports System.Data
Imports System.Windows.Forms
Imports System.Collections.Generic
Imports System.Text
Imports System.Linq

Public Class Form1
    Inherits Form

#Region " Definición de los controles y Sub Main() "

    Public Sub New()
        MyBase.New()

        'This call is required by the Windows Form Designer.
        InitializeComponent()

        'TODO: Add any initialization after the InitializeComponent call

    End Sub

    Shared Sub Main()
        Application.EnableVisualStyles()
        Application.SetCompatibleTextRenderingDefault(False)
        Application.Run(New Form1())
    End Sub

    '    'Form overrides dispose to clean up the component list.
    '    Public Overrides Sub Dispose()
    '        MyBase.Dispose()
    '        If Not (components Is Nothing) Then
    '            components.Dispose()
    '        End If
    '    End Sub


    'Required by the Windows Form Designer
    Private components As System.ComponentModel.Container

    'NOTE: The following procedure is required by the Windows Form Designer
    'It can be modified using the Windows Form Designer.  
    'Do not modify it using the code editor.
    Private Sub InitializeComponent()
        Me.txtTexto = New System.Windows.Forms.TextBox()
        Me.labelInfo = New System.Windows.Forms.Label()
        Me.btnAceptar = New System.Windows.Forms.Button()
        Me.btnCancelar = New System.Windows.Forms.Button()
        Me.SuspendLayout()
        '
        'txtTexto
        '
        Me.txtTexto.Anchor = CType(((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Left) _
            Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
        Me.txtTexto.ForeColor = System.Drawing.SystemColors.GrayText
        Me.txtTexto.Location = New System.Drawing.Point(12, 14)
        Me.txtTexto.Name = "txtTexto"
        Me.txtTexto.Size = New System.Drawing.Size(295, 20)
        Me.txtTexto.TabIndex = 0
        '
        'labelInfo
        '
        Me.labelInfo.AutoSize = True
        Me.labelInfo.Location = New System.Drawing.Point(12, 50)
        Me.labelInfo.Name = "labelInfo"
        Me.labelInfo.Size = New System.Drawing.Size(129, 13)
        Me.labelInfo.TabIndex = 3
        Me.labelInfo.Text = "<Escribe algo en el texto>"
        '
        'btnAceptar
        '
        Me.btnAceptar.Anchor = CType((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
        Me.btnAceptar.AutoSize = True
        Me.btnAceptar.Location = New System.Drawing.Point(313, 12)
        Me.btnAceptar.Name = "btnAceptar"
        Me.btnAceptar.Size = New System.Drawing.Size(75, 23)
        Me.btnAceptar.TabIndex = 1
        Me.btnAceptar.Text = "Saludar"
        '
        'btnCancelar
        '
        Me.btnCancelar.Anchor = CType((System.Windows.Forms.AnchorStyles.Bottom Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
        Me.btnCancelar.AutoSize = True
        Me.btnCancelar.DialogResult = System.Windows.Forms.DialogResult.Cancel
        Me.btnCancelar.Location = New System.Drawing.Point(313, 106)
        Me.btnCancelar.Name = "btnCancelar"
        Me.btnCancelar.Size = New System.Drawing.Size(75, 23)
        Me.btnCancelar.TabIndex = 2
        Me.btnCancelar.Text = "Cerrar"
        '
        'Form1
        '
        Me.AcceptButton = Me.btnAceptar
        Me.AutoScaleDimensions = New System.Drawing.SizeF(6.0!, 13.0!)
        Me.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font
        Me.CancelButton = Me.btnCancelar
        Me.ClientSize = New System.Drawing.Size(400, 141)
        Me.Controls.Add(Me.txtTexto)
        Me.Controls.Add(Me.btnAceptar)
        Me.Controls.Add(Me.btnCancelar)
        Me.Controls.Add(Me.labelInfo)
        Me.Name = "Form1"
        Me.Text = "Form1 VB"
        Me.ResumeLayout(False)
        Me.PerformLayout()

    End Sub

    '
    ' Declaración de los controles
    '
    Private WithEvents txtTexto As TextBox
    Private labelInfo As Label
    Private WithEvents btnAceptar As Button
    Private WithEvents btnCancelar As Button

#End Region

    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles Me.Load
        ' Centrar en la pantalla
        Me.CenterToScreen()

        ' Aceptar comprobaci n de teclas
        Me.KeyPreview = True

        txtTexto.Text = textVacio
        txtTexto.SelectionStart = 0 'txtTexto.Text.Length
    End Sub

    Private Sub Form1_Closing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing
        ' Cuando se cierra el formulario

    End Sub

    Private Sub Form1_KeyDown(sender As Object, e As KeyEventArgs) Handles Me.KeyDown
        ' Detectar teclas

    End Sub

    Private Sub btnAceptar_Click() Handles btnAceptar.Click
        If txtTexto.Text = textVacio Then
            labelInfo.Text = "¡Hola amigo!"
        Else
            labelInfo.Text = $"¡Hola {txtTexto.Text}!"
        End If
    End Sub

    Private Sub btnCancelar_Click() Handles btnCancelar.Click
        Me.Close()
    End Sub

    '
    ' Lo que necesitamos para hacer el efecto
    '

    Private Const textVacio As String = "<Escribe tu nombre>"
    Private textAnterior As String = textVacio
    Private inicializando As Boolean

    Private Sub txtTexto_Enter(sender As Object, e As EventArgs) Handles txtTexto.Enter
        If txtTexto.Text <> textVacio Then
            txtTexto.ForeColor = SystemColors.ControlText
        End If
    End Sub

    Private Sub txtTexto_Leave(sender As Object, e As EventArgs) Handles txtTexto.Leave
        If String.IsNullOrEmpty(Me.txtTexto.Text) Then '
            Me.txtTexto.ForeColor = SystemColors.GrayText
            Me.txtTexto.Text = textVacio
        End If
    End Sub

    Private Sub txtTexto_TextChanged(sender As Object,
                                           e As EventArgs) Handles txtTexto.TextChanged
        If inicializando Then Return
        If txtTexto.Text = "" OrElse txtTexto.Text = textVacio Then
            txtTexto.ForeColor = SystemColors.GrayText
            inicializando = True
            txtTexto.Text = textVacio
            inicializando = False
        Else
            If textAnterior = textVacio Then
                inicializando = True
                txtTexto.Text = QuitarPredeterminado(txtTexto.Text, textVacio)
                inicializando = False

                txtTexto.SelectionStart = txtTexto.Text.Length

            End If
            txtTexto.ForeColor = SystemColors.ControlText
        End If
        textAnterior = txtTexto.Text
    End Sub

    ''' <summary>
    ''' Quitar de una cadena un texto indicado (que será el predeterminado cuando está vacío).
    ''' Por ejemplo si el texto grisáceo es Buscar... y
    ''' se empezó a escribir en medio del texto (o en cualquier parte)
    ''' BuscarL... se quitará Buscar... y se dejará L.
    ''' Antes de hacer cambios se comprueba si el texto predeterminado está al completo 
    ''' en el texto en el que se hará el cambio.
    ''' </summary>
    ''' <param name="texto">El texto en el que se hará la sustitución.</param>
    ''' <param name="predeterminado">El texto a quitar.</param>
    ''' <returns>Una cadena con el texto predeterminado quitado.</returns>
    ''' <remarks>18/Oct/2020 actualizado 24/Oct/2020</remarks>
    Public Function QuitarPredeterminado(texto As String, predeterminado As String) As String
        Dim cuantos = predeterminado.Length
        Dim k = 0

        For i = 0 To predeterminado.Length - 1
            Dim j = texto.IndexOf(predeterminado(i))
            If j = -1 Then Continue For
            k += 1
        Next
        ' si k es distinto de cuantos es que no están todos lo caracteres a quitar
        If k <> cuantos Then
            Return texto
        End If

        For i = 0 To predeterminado.Length - 1
            Dim j = texto.IndexOf(predeterminado(i))
            If j = -1 Then Continue For
            If j = 0 Then
                texto = texto.Substring(j + 1)
            Else
                texto = texto.Substring(0, j) & texto.Substring(j + 1)
            End If
        Next

        Return texto
    End Function

End Class

